staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"fmt"
+	"net/netip"
+	"strings"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// CIDR provides a CEL function library extension of CIDR notation parsing functions.
+//
+// cidr
+//
+// Converts a string in CIDR notation to a network address representation or results in an error if the string is not a valid CIDR notation.
+// The CIDR must be an IPv4 or IPv6 subnet address with a mask.
+// Leading zeros in IPv4 address octets are not allowed.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4/24) are not allowed.
+//
+//	cidr(<string>) <CIDR>
+//
+// Examples:
+//
+//	cidr('192.168.0.0/16') // returns an IPv4 address with a CIDR mask
+//	cidr('::1/128') // returns an IPv6 address with a CIDR mask
+//	cidr('192.168.0.0/33') // error
+//	cidr('::1/129') // error
+//	cidr('192.168.0.1/16') // error, because there are non-0 bits after the prefix
+//
+// isCIDR
+//
+// Returns true if a string is a valid CIDR notation respresentation of a subnet with mask.
+// The CIDR must be an IPv4 or IPv6 subnet address with a mask.
+// Leading zeros in IPv4 address octets are not allowed.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4/24) are not allowed.
+//
+//	isCIDR(<string>) <bool>
+//
+// Examples:
+//
+//	isCIDR('192.168.0.0/16') // returns true
+//	isCIDR('::1/128') // returns true
+//	isCIDR('192.168.0.0/33') // returns false
+//	isCIDR('::1/129') // returns false
+//
+// containsIPAddr / containsCIDR / ip / masked
+//
+// - containsIPAddr: Returns true if a network address contains an IP address. Should be passed an IP address parsed with ipAddr.
+//
+// - containsCIDR: Returns true if a network address contains a CIDR. Should be passed a CIDR parsed with cidr.
+//
+// - ip: Returns the IP address representation of the CIDR.
+//
+// - masked: Returns the CIDR representation of the network address with a masked prefix.
```

Comment:
You may need a way to query how many prefix bits and/or host bits - ISTR a few places where we enforce maximum size, e.g.

`cidr.family() == 4 ? cidr.hostbits <= 16 : cidr.hostbits <= 64`

I'm not sure what the CIDR spec names those. :)

@aojea

Reply:
The number after the slash is the "prefix length". I'm not sure there's a standard name for the total length minus prefix length. "Host bits" only makes sense in the context of node cidrs, where it means "the number of bits reserved for this host to use for pod IPs" (but if you run `kcm --cluster-cidr 10.128.0.0/20`, it wouldn't really make sense to say that the cluster cidr has "12 host bits", because 4 of those bits are used in one way and 8 in another way).

FWIW I'm not sure we actually care about validating that anywhere except for service network CIDRs, but if we're going to have a selector to get the IP-address part of a CIDR, then it makes sense to have a selector to get the prefix-length part too..

Reply:
So should we expose this under `prefixLength()` then? Would that be the most appropriate/expected name?

Reply:
sure, `prefixLength` sounds right